### PR TITLE
Feature/req beacon state

### PR DESCRIPTION
**Pull request by Cortze**
Aggregate the beacon.Status to Peer struct and include the needed functions to request the status from a given peer 
NOTE: I'm unsure of which is the best place where to aggregate the ReqBeaconStatus()

### CODE REVIEW

File: src/metrics/peer.go [hunk 1]
```diff
@@ -109,6 +115,15 @@ func (pm *Peer) GetAllMessagesCount() uint64 {
 		pm.ProposerSlashing.Cnt)
 }
 
+// Update beacon Status of the peer
+func (pm *Peer) UpdateBeaconStatus(bStatus beacon.Status) {
```
Comment (alrevuelta) [inline src/metrics/peer.go:119]: As I can see from [here](https://github.com/protolambda/zrnt/blob/9ef829e4a73f094061677cc6846f3dcdedb7a439/eth2/beacon/common/network.go#L235) you are storing the whole `beacon.Status`. Why not just:

```go
pm.BeaconStatus = bStatus
```

File: src/metrics/nodemetadata/metadata_req.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+		},
+		func(chunk reqresp.ChunkedResponseHandler) error {
+			resCode = chunk.ResultCode()
+			switch resCode {
+			case reqresp.ServerErrCode, reqresp.InvalidReqCode:
+				msg, err := chunk.ReadErrMsg()
+				if err != nil {
+					return err
+				}
+				return errors.New(fmt.Sprintf("error requesting BeaconState RPC: %s", msg))
```
Comment (alrevuelta) [inline src/metrics/nodemetadata/metadata_req.go:36]: can we wrap all errors so that we get the whole trace? `return errors.Wrap(err, "new error")`

File: src/metrics/nodemetadata/metadata_req.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
+package nodemetadata
```
Comment (alrevuelta) [inline src/metrics/nodemetadata/metadata_req.go:1]: what do you think about moving this outside of `metrics`? imho we should keep in metrics just the code that is not related to rumor. So that if in the future we use libp2p directly we can reuse everything as it is without any changes.

File: src/control/actor/host/notify.go [hunk 4]
```diff
@@ -60,14 +60,28 @@ func (c *HostNotifyCmd) connectedF(net network.Network, conn network.Conn) {
...
 	// request metadata as soon as we connect to a peer
 	peerData, err := PollPeerMetadata(conn.RemotePeer(), c.Base, c.PeerMetadataState, c.Store, c.PeerStore)
+	// request BeaconStatus metadata as we connect to a peer
+	h, err := c.Host()
+	if err != nil {
+		log.Error(err)
+	}
+	bState, err2 := nodemetadata.ReqBeaconStatus(context.Background(), h, conn.RemotePeer())
+	if err2 != nil {
+		log.Warn(err)
```
Comment (alrevuelta) [inline src/control/actor/host/notify.go:70]: careful, you are logging `err` not `err2`

File: src/control/actor/host/notify.go [hunk 5]
```diff
@@ -60,14 +60,28 @@ func (c *HostNotifyCmd) connectedF(net network.Network, conn network.Conn) {
 
 	// request metadata as soon as we connect to a peer
 	peerData, err := PollPeerMetadata(conn.RemotePeer(), c.Base, c.PeerMetadataState, c.Store, c.PeerStore)
+	// request BeaconStatus metadata as we connect to a peer
+	h, err := c.Host()
+	if err != nil {
+		log.Error(err)
+	}
+	bState, err2 := nodemetadata.ReqBeaconStatus(context.Background(), h, conn.RemotePeer())
```
Comment (alrevuelta) [inline src/control/actor/host/notify.go:68]: you can also use `err`

Comment (alrevuelta) [reply src/control/actor/host/notify.go:68]: Be careful when you continue after an error. Lets say that `h, err := c.Host()` errors. In that case you wouldn't want to continue calling `ReqBeaconStatus` because the content of `h` will be most likely `nil`. So most likely you will get a nul pointer exception here when line 64 errors. The whole point of errors is not continuing if there was one, as nasty things can happen.

File: src/control/actor/host/notify.go [hunk 6]
```diff
@@ -60,14 +60,28 @@ func (c *HostNotifyCmd) connectedF(net network.Network, conn network.Conn) {
 
 	// request metadata as soon as we connect to a peer
```
Comment (alrevuelta) [inline src/control/actor/host/notify.go:61]: I think we should add timestamps of both the metadata and the beacon status. Now its not very clear when we got that information. This can go in another PR but though but since it quite related I would mind having it in this one.

File: src/control/actor/host/notify.go [hunk 7]
```diff
@@ -60,14 +60,28 @@ func (c *HostNotifyCmd) connectedF(net network.Network, conn network.Conn) {
...
+	// request BeaconStatus metadata as we connect to a peer
+	h, err := c.Host()
+	if err != nil {
+		log.Error(err)
+	}
+	bState, err2 := nodemetadata.ReqBeaconStatus(context.Background(), h, conn.RemotePeer())
+	if err2 != nil {
+		log.Warn(err)
+	}
+	logrus.Info(&bState)
```
Comment (alrevuelta) [inline src/control/actor/host/notify.go:72]: Lets add logs a bit more meaningful, containing more information. Rebase to the latest #15 and you will see some examples here in `notify.go`

File: src/control/actor/host/notify.go [hunk 8]
```diff
@@ -60,14 +60,28 @@ func (c *HostNotifyCmd) connectedF(net network.Network, conn network.Conn) {
...
 	var peer metrics.Peer
+	// TODO: so far, If we didn't manage to exchange metadata, we asume that the peer didn't exchange Status neither
+	// 		  IMPORTANT: Metadata doen't include any kind of info related to the Host/Node
+	// 				     It just includes SeqNumber, Attnets
 	if err == nil {
-		peer = fetchPeerExtraInfo(peerData)
-		_ = peer
+		peer = fetchPeerExtraInfo(peerData, bState)
```
Comment (alrevuelta) [inline src/control/actor/host/notify.go:78]: I would mix the beacon state with the metadata. We could add two more methods to the peer.
* UpdateMetadata
* UpdateBeaconStatus

Each funtion updates the metadata/beaconstatus of the peer and also creates a time stamp on when it was fetched.

**Review by alrevuelta — CHANGES_REQUESTED** [review]
Great work! Looking forward to see the beacon status in action, really cool feature :) Left some comments.

File: src/metrics/peer.go (hunk 1 above)
Comment (Cortze) [reply src/metrics/peer.go:121]: Yep, much easier, I would like to add some middle struct with the time-stamp, so that I can implement this version tracker.

File: src/metrics/nodemetadata/metadata_req.go (hunk 3 above)
Comment (Cortze) [reply src/metrics/nodemetadata/metadata_req.go]: Yep, was a temporary place, moving it to the actor/host package

File: src/control/actor/host/notify.go (hunk 5 above)
Comment (Cortze) [reply src/control/actor/host/notify.go]: there is actually not much sense in tracking that error, if there is no host, it is impossible to get notified from a message and arrive at that function. I will dismiss the error making the sequence cleaner.

File: src/control/actor/host/notify.go [hunk 9]
```diff
@@ -57,27 +53,41 @@ func (c *HostNotifyCmd) listenCloseF(net network.Network, addr ma.Multiaddr) {
...
 func (c *HostNotifyCmd) connectedF(net network.Network, conn network.Conn) {
 	logrus.Info("connection detected: ", conn.RemotePeer().String())
-
-	// request metadata as soon as we connect to a peer
-	peerData, err := PollPeerMetadata(conn.RemotePeer(), c.Base, c.PeerMetadataState, c.Store, c.PeerStore)
+	h, _ := c.Host()
+	// Request the Host Metadata
+	hInfo := ReqHostInfo(context.Background(), h, conn)
+	// Request the BeaconMetadata
+	bMetadata, err := ReqBeaconMetadata(context.Background(), h, conn.RemotePeer())
```
Comment (alrevuelta) [inline src/control/actor/host/notify.go:60]: i think this error should be handled in some way. if it errors you will `UpdateBeaconMetadata` in line 82 with some wrong data

File: src/control/actor/host/notify.go [hunk 10]
```diff
@@ -57,27 +53,41 @@ func (c *HostNotifyCmd) listenCloseF(net network.Network, addr ma.Multiaddr) {
...
+	if err != nil {
+		log.Warn(err)
+	}
+	// request BeaconStatus metadata as we connect to a peer
+	bStatus, err := ReqBeaconStatus(context.Background(), h, conn.RemotePeer())
```
Comment (alrevuelta) [inline src/control/actor/host/notify.go:65]: same, as in 60. the error is not handled so `UpdateBeaconStatus` will be called with wrong information

File: src/control/actor/host/notify.go [hunk 11]
```diff
@@ -57,27 +53,41 @@ func (c *HostNotifyCmd) listenCloseF(net network.Network, addr ma.Multiaddr) {
...
-			PeerId: conn.RemotePeer().String(),
-		}
+	// Read ENR of the Peer from the generated enode
+	n := c.Store.LatestENR(conn.RemotePeer())
+	if n.ID().String() == "" {
+		// TODO: If the peer wasn't discovered via dv5 "n" will be empty
+		// we could compose new ENR with the beacon metadata and the host info
+		log.Warn("Peer ENR not found")
 	}
+	hInfo.NodeID = n.ID().String()
```
Comment (alrevuelta) [inline src/control/actor/host/notify.go:77]: as you are saying above, "If the peer wasn't discovered via dv5 "n" will be empty". here you are storing it regardless. perhaps move this check inside the `if`?
but not sure i understand it: is there any other way of getting a peer that is not via dv5?

File: src/control/actor/host/notify.go [hunk 12]
```diff
@@ -57,27 +53,41 @@ func (c *HostNotifyCmd) listenCloseF(net network.Network, addr ma.Multiaddr) {
...
+	// fetch all the info gathered from the peer into a new Peer struct
+	peer.FetchHostInfo(hInfo)
+	// TODO: Switch from Update ot Add in the future, when implementing metrics versions
+	peer.UpdateBeaconStatus(bStatus)
+	peer.UpdateBeaconMetadata(bMetadata)
+	// So far, just act like if we got new info, Update or Aggregate new info from a peer already on the Peerstore
 	c.PeerStore.AddPeer(peer)
 	c.PeerStore.AddConnectionEvent(conn.RemotePeer().String(), "Connection")
```
Comment (alrevuelta) [inline src/control/actor/host/notify.go:85]: small nitpick, but now that i see it. We can also directly call `peer.ConnectionEvent` before adding the peer to the `peerstore`. Cleaner imho:

```
peer.FetchHostInfo(hInfo)
peer.UpdateBeaconStatus(bStatus)
peer.UpdateBeaconMetadata(bMetadata)
peer.ConnectionEvent(x, y)
```

File: src/control/actor/host/req_metadata.go [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+		})
+	return
+}
+
+type HostWithIDService interface {
+	IDService() *identify.IDService
+}
+
+// ReqHostInfo request the host infomartion regarding a given peer, from the libp2p perspective
+// return empty struct and error if failure
```
Comment (alrevuelta) [inline src/control/actor/host/req_metadata.go:97]: perhaps its better to actually return an error if there is a failure.

File: src/control/actor/host/req_metadata.go [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+		},
+		func(chunk reqresp.ChunkedResponseHandler) error {
+			resCode = chunk.ResultCode()
+			switch resCode {
+			case reqresp.ServerErrCode, reqresp.InvalidReqCode:
+				msg, err := chunk.ReadErrMsg()
+				if err != nil {
+					return err
+				}
+				return errors.Errorf("error requesting BeaconState RPC: %s", msg)
```
Comment (alrevuelta) [inline src/control/actor/host/req_metadata.go:46]: wrap the error that we get from above with extra information
`errors.Wrap(err, "some new info")`

File: src/control/actor/host/req_metadata.go [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+		})
+	return
+}
+
+func ReqBeaconMetadata(ctx context.Context, h host.Host, peerID peer.ID) (data beacon.MetaData, err error) {
+	// Generate the compression
+	comp := reqresp.SnappyCompression{}
+	// Generate the Server Error Code
+	var resCode reqresp.ResponseCode // error by default
+	err = methods.MetaDataRPCv1.RunRequest(ctx, h.NewStream, peerID, comp, reqresp.RequestSSZInput{Obj: &data}, 1,
```
Comment (alrevuelta) [inline src/control/actor/host/req_metadata.go:66]: all this seems a bit difficult to read to me, don't you think so?

File: src/metrics/peer.go [hunk 16]
```diff
@@ -72,6 +80,9 @@ func NewPeer(peerId string) Peer {
 		MetadataRequest: false,
 		MetadataSucceed: false,
 
+		BeaconStatus:   NewBStatusStamped(beacon.Status{}),
```
Comment (alrevuelta) [inline src/metrics/peer.go:83]: cant we just have `beacon.Status{}` ?

File: src/metrics/peer.go [hunk 17]
```diff
@@ -109,6 +120,75 @@ func (pm *Peer) GetAllMessagesCount() uint64 {
 		pm.ProposerSlashing.Cnt)
 }
 
+// Fetch all the different metadata we received to save it into a new peer struct
+// TODO: Still few things to consider with new approach, like version handling
+// 		 or fetching the actual info with previous info from the peer
+func (pm *Peer) FetchHostInfo(hInfo BasicHostInfo) {
```
Comment (alrevuelta) [inline src/metrics/peer.go:126]: not sure i understand this function. imho `peer.go` should be as dummy as possible and i have the feeling it adds some very specific things

File: src/metrics/remote_peer_info.go [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+	Direction string
+	// Metadata requested
+	MetadataRequest bool
+	MetadataSucceed bool
+}
+
+// BEACON METADATA
+
+// Basic BeaconMetadata struct that includes the timestamp of the received beacon metadata
+type BeaconMetadataStamped struct {
```
Comment (alrevuelta) [inline src/metrics/remote_peer_info.go:33]: shouldnt this be part of the peer.go?

File: src/metrics/remote_peer_info.go [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+	Timestamp time.Time
+	Metadata  beacon.MetaData
+}
+
+// Generate a New BeaconMetadata struct and timestamp it
+// NOTE: It will append empty metadatas timestamped to track when did we request them
+func NewBMetadataStamped(meta beacon.MetaData) BeaconMetadataStamped {
```
Comment (alrevuelta) [inline src/metrics/remote_peer_info.go:40]: shouldnt this be part of the peer.go?
perhaps we dont need a constructor for this, as its easier to just `BeaconMetadataStamped{t.Now(), meta)`

File: src/metrics/utils/utils.go [hunk 20]
```diff
@@ -78,3 +81,36 @@ func GetTimeMiliseconds() int64 {
 
 	return millis
 }
+
+func GetIPfromMultiaddress(multiaddr string) (ip string, err error) {
```
Comment (alrevuelta) [inline src/metrics/utils/utils.go:85]: do we still need this?

**Review by alrevuelta — CHANGES_REQUESTED** [review]
great work! left some comments. I think you need to rebase it to the latest refactoring (with your stuff on top merged) #15

File: src/control/actor/host/notify.go (hunk 9 above)
Comment (Cortze) [reply src/control/actor/host/notify.go:60]: In this case, ReqBeaconMetadata will always provide an empty struct of the Beacon.MetaData. 
In case of not receiving it, I thought that a warning log was more than enough since we wanted to implement a data version struct in the future.

Do you think that just fetching the info to the Peer struct should be only done if the `err != nil` ?

Perhaps, so far, we would just like to update the Peer info if the request was successful, what do you think?

File: src/control/actor/host/notify.go (hunk 11 above)
Comment (Cortze) [reply src/control/actor/host/notify.go:77]: you are right, the assignment of the nodeID should be only done if it was in our dv5 database. 
It can be possible (remote possibility but still feasible) that a peer gets connected for its initiative when we didn't receive info about it in dv5 before. 
I'm not aware of any method to obtain the nodeID without the ENR if it's not by having the privkey. Therefore, we would have to wait until we discover that peer in dv5 to later obtain the nodeID and update it. 
I would like to know your thoughts about this, seems like a very rare case though.

File: src/metrics/utils/utils.go (hunk 20 above)
Comment (Cortze) [reply src/metrics/utils/utils.go:85]: As we saw, there are some IPs shared on the ENR that belong to a restringed or a private range, I would still use this function to extract the real IP from the established connection

File: src/metrics/remote_peer_info.go (hunk 18 above)
Comment (Cortze) [reply src/metrics/remote_peer_info.go:33]: I don't have any strong opinion about having it on the same peer.go file, I thought that it could look cleaner separating it. If you think that it belongs to the peer.go I can move it there.

File: src/metrics/peer.go (hunk 17 above)
Comment (Cortze) [reply src/metrics/peer.go:126]: I wanted to discuss this implementation with you. I was perhaps going further than I should, but my idea was to prepare the function and the metadata to track possible versions in the future. This struct also summarizes all the information that we can stract from a remote peer by using libp2p identify requests.
This way, we could track this basic versioned info about each remote hosts. Perhaps all this libp2p host-related info could be stored in the Peer struct directly

File: src/control/actor/host/req_metadata.go (hunk 13 above)
Comment (Cortze) [reply src/control/actor/host/req_metadata.go:97]: I will add an error flag in case the libp2p identify request fails so we can also log it.

File: src/metrics/remote_peer_info.go (hunk 19 above)
Comment (Cortze) [reply src/metrics/remote_peer_info.go:40]: I will remove the constructor yep, looks cleaner.

File: src/control/actor/host/notify.go (hunk 9 above)
Comment (alrevuelta) [reply src/control/actor/host/notify.go:63]: What I mean is that if you don't check that `ReqBeaconMetadata` gave an error, you will end up calling `UpdateBeaconMetadata` with an empty `bMetadata`. Is that what you want?
```
peer.UpdateBeaconMetadata(bMetadata)
```

So what I'm suggesting is doing:
```
if err != nil {
  peer.UpdateBeaconMetadata(bMetadata)
}
```

File: src/metrics/utils/utils.go (hunk 20 above)
Comment (alrevuelta) [reply src/metrics/utils/utils.go:85]: out of curiosity, have you seen any enr containing an ip in the range `192.168.x.x` for example? I remember we saw this 100.101... but have we seen any of the typical private network ones? i doubt it but just curious.

File: src/control/actor/host/notify.go (hunk 9 above)
Comment (Cortze) [reply src/control/actor/host/notify.go:71]: Updated in latest version

File: src/metrics/utils/utils.go (hunk 20 above)
Comment (Cortze) [reply src/metrics/utils/utils.go:100]: Noup, I didn't see so far any ENR returning private networks such as 127.0... or 192.168... , I just saw restringed IPs

File: src/control/actor/host/req_metadata.go [hunk 21]
```diff
@@ -0,0 +1,172 @@
...
+	}
+	// Fulfill the hInfo struct
+	ua, err := h.Peerstore().Get(peerID, "AgentVersion")
+	if err == nil {
+		peer.UserAgent = ua.(string)
+	} else {
+		// EDGY CASE: when peers refuse the connection, the callback gets called and the identify protocol
+		// returns an empty struct (we are unable to identify them)
+		err_ident = errors.Errorf("identification error caused by connection refuse")
+		peer.MetadataSucceed = false
```
Comment (alrevuelta) [inline src/control/actor/host/req_metadata.go:162]: perhaps this flag should be set in other place (notify.go?), after actually getting an error when requesting the metadata.

File: src/metrics/peer.go [hunk 22]
```diff
@@ -95,6 +100,76 @@ func (pm *Peer) ConnectionAttemptEvent(succeed bool, err string) {
...
+	}
+	// Beacon Metadata and Status
+	if newPeer.BeaconMetadata != (BeaconMetadataStamped{}) {
+		pm.BeaconMetadata = newPeer.BeaconMetadata
+	}
+	if newPeer.BeaconStatus != (BeaconStatusStamped{}) {
+		pm.BeaconStatus = newPeer.BeaconStatus
+	}
+	// Aggregate connections and disconnections
+	for _, time := range newPeer.ConnectionTimes {
```
Comment (alrevuelta) [inline src/metrics/peer.go:140]: Not sure why this. Why not just this without the loop?

```go
pm.ConnectionEvent(newPeer.ConnectedDirection, time)
```

File: src/metrics/peer.go [hunk 23]
```diff
@@ -95,6 +100,76 @@ func (pm *Peer) ConnectionAttemptEvent(succeed bool, err string) {
...
+		pm.ConnectionEvent(newPeer.ConnectedDirection, time)
+	}
+	for _, time := range newPeer.DisconnectionTimes {
```
Comment (alrevuelta) [inline src/metrics/peer.go:143]: same

**Review by alrevuelta — CHANGES_REQUESTED** [review]
Great! Some minor stuff and we are ready to merge

File: src/metrics/peer.go (hunk 22 above)
Comment (Cortze) [reply src/metrics/peer.go:140]: what if we want to make a fetch info from a peer that had more than 1 connection, this way either if there are no connections, either there is more than 1, it will include any slice of the ConnectedTimes[] to the peer where we are fetching the info.
Does this make sense?

File: src/control/actor/host/req_metadata.go (hunk 21 above)
Comment (Cortze) [reply src/control/actor/host/req_metadata.go:162]: Can look cleaner yep

File: analyzer/armiarma-analyzer.py [hunk 24]
```diff
@@ -630,12 +630,12 @@ def main():
 	        totalMessageCounter = totalMessageCounter + row['Total Messages']
```
[the remark this answers is not in the document]
Comment (Cortze) [reply analyzer/armiarma-analyzer.py:630]: 100% with you here. It is definitely something that should be on our TODO list. At the same time, we currently focusing on the web-realtime dashboard, maybe not that worth spending time refactoring it (maybe a quick refresh could be interesting). Let's take a look at it in few weeks.

File: src/metrics/peer.go (hunk 22 above)
Comment (alrevuelta) [reply src/metrics/peer.go:140]: Yep, sorry, I wasnt realising that the input was another peer, and of course as you are saying that other peer can have also other connection times. perhaps we can call it `MergePeer` or something like that.

**Review by alrevuelta — APPROVED** [review]
great work!
